src/dune/exe_rules.ml

```diff
@@ -91,18 +91,47 @@ let executables_rules ~sctx ~dir ~expander ~dir_contents ~scope ~compile_info
      make the code more uniform. *)
   let ext_lib = ctx.lib_config.ext_lib in
   let link_args =
+    let paths_to_relative_foreign_archives () =
```

Comment:
Why is this code (collection of .a archives) is separate from the other code that collects .a archives for linking? Don't their jobs overlap greatly?

I dug a little and I think I found the relevant code:

- `Exe.build_and_link_many` calls:
- `Link_time_code_gen.handle_special_libs` constructs:
- `Lib.Lib_and_module.Lib `

which is later passed via `link_time_code_gen.to_link` to:
- `link_exe`, then to `Lib.Lib_and_module.L.link_flags`

So I guess my question is: 
- why is this distinct from `Lib.Lib_and_module.L.link_flags`?
- can we try to simplify and unify these things? (the current data flow seems convoluted almost as if on purpose)

Reply:
Indeed, computation of link flags is quite convoluted.

> why is this distinct from `Lib.Lib_and_module.L.link_flags`?

This PR extends an existing mechanism of passing `-cclib` flags for "direct foreign archives" by passing flags for "transitive foreign archives" too. I guess your question applies equally to the existing code and the new extension, and I think the answer is that `link_time_code_gen` seems to be related to special libraries; indeed, it can only be created via `handle_special_libs`.

> can we try to simplify and unify these things? 

I am sure we can simplify and unify this. I wonder if this is better done in a separate PR?

Reply:
Talked about this in person. Andrey is going to try to fix existing code because it looks like an issue in `Lib.link_flags` and `Lib.include_flags` (really terribly named when used for linker args). Doesn't seem worth adding so much new code to paper over bugs in existing code.

Reply:
Done. The result is much clearer indeed!